Simplify process (#299)
waku-org/pm: PROCESS.md

- Individual FURPS (e.g a specific `F`unctionality statement) should be owned by one and only one Waku subteam.
- "a FURPS" refers to an individual FURPS statement (e.g. "S1. available in libwaku").
- For every `P` (Performance) there must be a Grafana panel (pointed to fleet), and a **Vac-DST** simulation to *sign-off* the `P` (search for “**Vac-DST**”).
- For every `R` (Reliability) there should be a test suite by Vac-QA that sign-off the `R` in unreliable network environment (search for “**Vac-QA**”);  and potentially a Grafana panel (pointed to fleet), and a Vac-DST simulation (if relevant).

## Feature

A feature is a specific domain of Waku software behaviour.
It can be a protocol, or a set of closely-related protocols.
It has one related set of FURPS.

For example, "End-to-end reliability" is a feature that refers to the SDS protocol.

## Milestone

A *Milestone* define a completion date and effort estimate for a collection of FURPS within or across given features.
Milestones are set for the whole Waku team and may involve the effort of one or several Waku subteams.
Dependencies on other groups to achieve a milestone should be minimalized.

> ## Deliverable

Author: We could also remove "deliverable" and abuse the "+" of [FURPS+](https://www.marcinziemek.com/blog/content/articles/8/article_en.html)

For example

- +1. CONTRIBUTOR.md is created in nim-sds repo
- +2. Video demo of SDS in Qaku is published

- Individual FURPS (e.g a specific `F`unctionality statement) should be owned by one and only one Waku subteam.
- "a FURPS" refers to an individual FURPS statement (e.g. "S1. available in libwaku").
- For every `P` (Performance) there must be a Grafana panel (pointed to fleet), and a **Vac-DST** simulation to *sign-off* the `P` (search for “**Vac-DST**”).
- For every `R` (Reliability) there should be a test suite by Vac-QA that sign-off the `R` in unreliable network environment (search for “**Vac-QA**”);  and potentially a Grafana panel (pointed to fleet), and a Vac-DST simulation (if relevant).

### Feature

A feature is a specific domain of Waku software behaviour.
It can be a protocol, or a set of closely-related protocols.
It has one related set of FURPS.

For example, "End-to-end reliability" is a feature that refers to the SDS protocol.

### Milestone

A *Milestone* define a completion date and effort estimate for a collection of FURPS within or across given features.
Milestones are set for the whole Waku team and may involve the effort of one or several Waku subteams.
Dependencies on other groups to achieve a milestone should be minimalized.

> ### Deliverable

Author: We could also remove "deliverable" and abuse the "+" of [FURPS+](https://www.marcinziemek.com/blog/content/articles/8/article_en.html)

For example

- +1. CONTRIBUTOR.md is created in nim-sds repo
- +2. Video demo of SDS in Qaku is published

  - What work is planned
  - What work is in progress
  - What work is done

Software properties are considered *delivered* when:

- it is ready-to-use
- the behaviour is documented in [specifications](https://waku.org/specs)
- the software has been used (dogfooding) by the Waku team

See [Waku Methodology](https://www.notion.so/Waku-Mission-and-Methodology-1a58f96fb65c80b9b789c1bbb9e99915) (draft) for more.

## Team Structure

The Waku team is currently organized in the following subteams:

- Core Research
- nwaku
- js-waku
- Chat/App Dev
- Chat/App Research
- Business Dev

## Documents

- [Waku FURPS](https://www.notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00)
- [Waku Milestones](https://roadmap.logos.co/waku/waku-milestones)

## Terminology and Scope

### FURPS

[FURPS](https://en.wikipedia.org/wiki/FURPS) are used to define software behaviour and property.
They are defined as:
- F: Functionality
- U: Usability
- R: Reliability
- P: Performance
- S: Supportability

Moreover: 

- A *set of FURPS* define the behaviour of a specific protocol implementation or feature.
- "a FURPS" refers to an individual FURPS statement (e.g. "S1. available in libwaku").

> - For every `P` (Performance) there must be a Grafana panel (pointed to fleet), and a **Vac-DST** simulation to *sign-off* the `P` (search for “**Vac-DST**”).

Reviewer: Does this imply  "if there cannot be a Grafana Panel then the requirement ought not be listed under `P`"

Author: I agree that if you can't measure it, you can't state it.
But Grafana may not be the breaking point here, as it might possible to measure it in DST simulation without a dedicated Prometheus metric. Will update in any case.

## Motivation and Goal

Provide a clear visibility to anyone (CCs or not) of:
  - What work is planned
  - What work is in progress
  - What work is done

Software properties are considered *delivered* when:

- it is ready-to-use
- the behaviour is documented in [specifications](https://waku.org/specs)
- the software has been used (dogfooding) by the Waku team
- documentation on how to use said properties is delivered

See [Waku Methodology](https://www.notion.so/Waku-Mission-and-Methodology-1a58f96fb65c80b9b789c1bbb9e99915) (draft) for more.

## Team Structure

The Waku team is currently organized in the following subteams:

- Core Research
- nwaku
- js-waku
- Chat/App Dev
- Chat/App Research
- Business Dev

## Documents

> - [Waku FURPS](https://www.notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00)

Reviewer 2: nit: should this be published to `roadmap.logos.co` too? 
milestones and FURPS are intervened

Reviewer 3: I agree since the roadmap.logos site intends to "inform the previous, current, and future work required to fulfill the requirements of the projects under the Logos Collective", we might as well include FURPS definitions in a section there, nested under Milestones perhaps.

Author: Good idea. I think for this time around I'd prefer to iterate in this repo/PR: https://github.com/waku-org/pm/pull/301 but happy to move it to logos roadmap repo once finalized and review if we can iterate directly there nex time.

Reviewer 4: Great idea!

Besides, I think we should move the FURPS definitions into the pm repository as Wiki pages.

We have a great Milestone → Deliverable → Issue tree structure on GitHub, but there is no straightforward connection between that structure and FURPS. In other words, it's easy to navigate from FURPS to a Milestone, but not the other way around.

My suggestion:

  - Create a FURPS-Wiki page for each milestone. On that FURPS-Wiki page we can also add links to the related milestone and deliverables.
  - Each Milestone's definition can contain a link to the FURPS-Wiki page within the pm repo.

> See [Waku Methodology](https://www.notion.so/Waku-Mission-and-Methodology-1a58f96fb65c80b9b789c1bbb9e99915) (draft) for more.

Author: I suggest we move the Waku methodology to a **good place** (Waku website?) once settled.
We can iterate on it as part of virtual offsite.

> See [Waku Methodology](https://www.notion.so/Waku-Mission-and-Methodology-1a58f96fb65c80b9b789c1bbb9e99915) (draft) for more.

Author: Also note that I discourage the usage of GitHub wiki pages. It is it not backed up on Git Tea and just creates "one more location" to maintain or keep track of.

- [Waku Milestones](https://roadmap.logos.co/waku/waku-milestones)

## Terminology and Scope

### FURPS

[FURPS](https://en.wikipedia.org/wiki/FURPS) are used to define software behaviour and property.
They are defined as:
- F: Functionality
- U: Usability
- R: Reliability
- P: Performance
- S: Supportability

Moreover: 

- A *set of FURPS* define the behaviour of a specific protocol implementation or feature.
- "a FURPS" refers to an individual FURPS statement (e.g. "S1. available in libwaku").
- For every `P` (Performance) there should be a Grafana panel (pointed to fleet), and a **Vac-DST** simulation to *sign-off* the `P` (search for “**Vac-DST**”).
  If it cannot be measured, then it cannot be a statement (in some instances, DST simulation may be enough).
- For every `R` (Reliability) there should be a test suite by Vac-QA that sign-off the `R` in unreliable network environment (search for “**Vac-QA**”);  and potentially a Grafana panel (pointed to fleet), and a Vac-DST simulation (if relevant).

> ### Feature

Reviewer 4: tbh , I wouldn't add another term :)
I think is enough with milestone , deliverable , task , FURPS :melting_face:

Author: Not really a new term per se, the Waku software is already split in various functional domain aka protocols aka features.

Author: Another way to think about it is that "FURPS per milestone" do not work. For example:

-  Milestone e2e reliability defines FURPS for SDS to be implemented in Nim
- Then we do a Milestone e2e reliability for the browser, to implement SDS in JS.

If we have FURPS per milestone, then we would need to redefine the SDS FURPS for the browser milestone. Which now creates some duplicate work.

FURPS describe software behaviour and their properties.

A milestone tracks the implementation of specific FURPS.

Hence, "FURPS per milestone" do not make sense or work.

Reviewer 4: I thought there was 1:1 Milestone<->FURPS-set ratio because we have one milestone link on every FURPS set: https://www.notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00

Also, with regards to the `e2e reliability protocol` example, maybe we can have separate Deliverables per platform (js-waku; nwaku) similarly to what we have in `RLN mainnet` milestone.

Not a big deal though and thanks for clarifying!

Author: > I thought there was 1:1 Milestone<->FURPS-set ratio because we have one milestone link on every FURPS set: [notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00](https://www.notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00)

We are moving away from this model because a milestone describe an enhancement to be done by a date.

FURPS are timeless, they describe the properties of the software.

So the relation is:
- 1 milestone is a checkpoint we are aiming for Waku. Such as "stabilize TWN" (dummy example).
- To achieve this milestone, we may need to add N FURPS to various protocols and functions, for example:
  - improve bootstrap: peer exchange, rendezvous enhancements (peer exchange FURPS, rendezvous FURPS)
  - fix web: yamux, websocket, webtransport (js-waku FURPS, nwaku FURPS)
  - Fix nwaku memory: nwaku (nwaku FURPS)
- To organize the work and express clear accountability, we define deliverables:
  - "Improve nwaku as a boostrap node": owned by nwaku team, delivers new FURPS on rendezvous, etc
  - "Improve js-waku connectivity at start":  owned by js-waku team, delivers new js-waku FURPS on js-waku etc
  -  "Fix memory in nwaku": owned by nwaku, delibers new nwaku FURPS
  
  > Also, with regards to the e2e reliability protocol example, maybe we can have separate Deliverables per platform (js-waku; nwaku) similarly to what we have in RLN mainnet milestone.
 
Yes, each deliverable is owned by one team, and hence one platform.

Author: ![signal-2025-05-28-150859](https://github.com/user-attachments/assets/caa59f3b-28ab-4036-9dc2-58597fe87a68)

Here is a poor man diagram

A feature is a specific domain of Waku software behaviour.
It can be a protocol, or a set of closely-related protocols.
It has one related set of FURPS.

For example, "End-to-end reliability" is a feature that refers to the SDS protocol.

### Deliverable

A deliverable is the work tackled by one subteam in regards to one FURPS set.

> In other words, it tracks the delivery of specific FURP(S), for one feature, by one subteam.

Reviewer 4: ```suggestion
In other words, it tracks the delivery of specific FURP(S) by one or many subteams.
```

Author: I want to remove ownership confusion so a deliverable will belong to a unique team.

For example:

> Light Push FURPS
> - F1. Light push returns itemized error code when the service node cannot forward a message.
> - S1. nwaku service node
> - S2. Browser edge node

To track the work to achieve these FURPS, two deliverables should be created: 

1. [nwaku] Support error code in light push: Implement Light Push FURPS F1, S1.
2. [js-waku] Support error code in light push: Implement Light Push FURPS F1, S2.

When the **work** is not related to software behaviour (e.g. a contributor guide, some BD activities),
then a deliverable is used to define the expected output, without referring to specific FURPS.

Deliverables are owned by one, and only one, subteam.
If a deliverable's execution depends on a group external to Waku, then it should be explicitly stated.

### Milestone

A milestone describes a specific objective we aim to achieve.
It holds the narrative of **why** we are doing this work and what is the value attached to completing the work.

> A milestone usually involve effort from most of the Waku team.

Reviewer 4: ```suggestion
A milestone usually involves effort from most of the Waku team.
```

- MUST have a GitHub Milestone in https://github.com/waku-org/pm repo, to which relevant _FURPS_ statements or _Deliverables_ are added.
- MUST have an *estimated date of completion*
- MUST have an effort estimate, stating how many CCs are needed to work on this for a given half-year (e.g. one research, half an engineer)

A _Deliverable_:
- MUST be defined as an issue in the https://github.com/waku-org/pm repo.
- MUST be included in its parent _Milestone_.
- MUST have an _Output_ section in the description detailing the result of work of the Deliverable, this may be a list of FURPS.
- If tracking FURPS, the FURPS only belong to one feature aka FURPS set.
- MUST have only one owner, assigned to the GitHub issue.
- MUST have a checklist to ensure the following items are done:
  - specs
  - code
  - dogfooding
  - docs

Finally, for _Tasks_ that do not belong to a _Deliverable_:
- MUST either qualify as (with related GitHub labels)
  - `bug` - bugs reported by users or discovered internally, SHOULD be linked back to a corresponding _FURPS_ and _Milestone_
  - `test` - maintaining and fixing broken tests, SHOULD ideally be linked back to a corresponding _FURPS_ and _Milestone_

>   - `release` - work related to releasing version upgrades.

Author: A release is about applying a `tag` and bundling your code
dependencies is about updating your `package.json` and upgrading dependencies.

>   - `release` - work related to releasing version upgrades.

Reviewer 2: Then I would write something like:
- `release – work associated with releasing new versions`
- `dependencies – work associated with updating dependency versions`

>   - `release` - work related to releasing version upgrades.

Reviewer 4: I think this is not needed and nowadays we can make sure that all the tasks have a Deliverable as we've defined the "Maintenance Deliverable". See https://github.com/waku-org/pm/issues/275

Author: @chair28980 WDYT? Is this maintenance deliverable really needed? Can't we just use labels?

  - `dependencies` - work related to releasing version upgrades.

### Responsibilities

| Task                                                      | Does it         | Ensure it's done |
|-----------------------------------------------------------|-----------------|------------------|
| Set Milestones, FURPS and Deliverables in master document | Waku Lead       | Insights         |
| Create GitHub milestones in pm repo                       | Program Manager | Waku Lead        |
| Create Deliverables issues in pm repo                     | Team Leads      | Program Manager  |
| Create issues, PR (tasks) and link them to **FURPS**      | Team Member     | Team Lead        |
| Close Deliverables                                        | Waku Lead       | Program Manager  |
| Handover to Vac-QA, Vac-DST                               | Team Lead       | Vac PoCs         |

> | Proceed with Dogfooding                                   | Team Lead       | Waku Lead        |

Reviewer 2: Would like to clarify how `Proceed with Dogfooding` will happen in this structure and what Team Lead role is in it?

Author: I updated the table, let me know what you think

Reviewer 2: Looks good, but I still would like to clarify how the process would happen.

Presently, a team member is creating an item for dogfooding and does this. 
When a team lead is responsible, would it be expected for a team lead to create dogfooding items (issues, requests in Notion) and make a team member drive dogfooding?

Author: Update and changed all to "deliverable owner" because I do not want to dictate how Waku subteams self-organized.
Either the team lead does it, or the team member that owns the deliverable. It's up to the Waku subteam.

Waku Lead: @fryorcraken
Program Manager: @chair28980

> Team Lead: @plopezlpz @Ivansete-status @jm-clius @weboko @jazzz

Reviewer 2: this block doesn't seem to be formatted well for some reason

![image](https://github.com/user-attachments/assets/1ebd2b38-af58-42f6-bb17-45ecc199ec98)

Author: thx! fixed
